File: sop-java/src/main/java/sop/operation/InlineSign.java
// SPDX-FileCopyrightText: 2022 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.operation;

import sop.Ready;
import sop.enums.InlineSignAs;
import sop.exception.SOPGPException;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;

public interface InlineSign extends AbstractSign<InlineSign> {

    /**
     * Sets the signature mode.
     * Note: This method has to be called before {@link #key(InputStream)} is called.
     *
     * @param mode signature mode
     * @return builder instance
     *
     * @throws sop.exception.SOPGPException.UnsupportedOption if this option is not supported
     */
    InlineSign mode(InlineSignAs mode)
            throws SOPGPException.UnsupportedOption;

    /**
     * Signs data.
     *
     * @param data input stream containing data
     * @return ready
     *
     * @throws IOException in case of an IO error
     * @throws sop.exception.SOPGPException.KeyIsProtected if at least one signing key cannot be unlocked
     * @throws sop.exception.SOPGPException.ExpectedText if text data was expected, but binary data was encountered
     */
    Ready data(InputStream data)
            throws IOException,
            SOPGPException.KeyIsProtected,
            SOPGPException.ExpectedText;

    /**
     * Signs data.
     *
     * @param data byte array containing data
     * @return ready
     *
     * @throws IOException in case of an IO error
     * @throws sop.exception.SOPGPException.KeyIsProtected if at least one signing key cannot be unlocked
     * @throws sop.exception.SOPGPException.ExpectedText if text data was expected, but binary data was encountered
     */
    default Ready data(byte[] data)
            throws IOException,
            SOPGPException.KeyIsProtected,
            SOPGPException.ExpectedText {
        return data(new ByteArrayInputStream(data));
    }
}

File: sop-java/src/main/kotlin/sop/operation/InlineSign.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.operation

import java.io.IOException
import java.io.InputStream
import sop.Ready
import sop.enums.InlineSignAs
import sop.exception.SOPGPException.*

interface InlineSign : AbstractSign<InlineSign> {

    /**
     * Sets the signature mode. Note: This method has to be called before [.key] is called.
     *
     * @param mode signature mode
     * @return builder instance
     * @throws UnsupportedOption if this option is not supported
     */
    @Throws(UnsupportedOption::class) fun mode(mode: InlineSignAs): InlineSign

    /**
     * Signs data.
     *
     * @param data input stream containing data
     * @return ready
     * @throws IOException in case of an IO error
     * @throws KeyIsProtected if at least one signing key cannot be unlocked
     * @throws ExpectedText if text data was expected, but binary data was encountered
     */
    @Throws(IOException::class, KeyIsProtected::class, ExpectedText::class)
    fun data(data: InputStream): Ready

    /**
     * Signs data.
     *
     * @param data byte array containing data
     * @return ready
     * @throws IOException in case of an IO error
     * @throws KeyIsProtected if at least one signing key cannot be unlocked
     * @throws ExpectedText if text data was expected, but binary data was encountered
     */
    @Throws(IOException::class, KeyIsProtected::class, ExpectedText::class)
    fun data(data: ByteArray): Ready = data(data.inputStream())
}
